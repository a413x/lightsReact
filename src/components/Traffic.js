import React, {Component} from "react";
import Light from './Light.js';
import { Link } from 'react-router-dom';

import Navbar from './Navbar.js';

import '../styles/traffic.css';

class Traffic extends Component {
  constructor(props) {
    super(props);
    this.colors = [
      {id:0,color:'red'},
      {id:1,color:'yellow'},
      {id:2,color:'green'},
    ];
    this.times= [5, 4, 7];
    this.state = {
      timeCount: 0,
    };

    this.mainCycle = this.mainCycle.bind(this);

    this.prevPath = '';
    this.goNext = this.goNext.bind(this);
    this.goTo = this.goTo.bind(this);

    this.incCount = this.incCount.bind(this);
    this.setTimeCount = this.setTimeCount.bind(this);

    this.onSetClick = this.onSetClick.bind(this);

    this.intervalId = null;
  }

  mainCycle(){
    this.intervalId = setInterval(() => {
      this.incCount();
    }, 1000);
  }

  componentDidMount() {
    this.setTimeCount(this.props.lightsState.indexOf(1));
  }
  componentDidUpdate(prevProps) {
    if (prevProps.location.pathname !== this.props.location.pathname) {
      this.prevPath = prevProps.location.pathname;
    }
  }
  componentWillUnmount() {
    clearInterval(this.intervalId);
  }

  incCount() {
    if (this.state.timeCount <= 0) {
      this.goNext();
    } else {
      this.setState({
        timeCount: this.state.timeCount - 1
      });
    }
  }
  setTimeCount(ind){
    this.setState({
      timeCount: this.times[ind]
    });
    clearInterval(this.intervalId);
    this.mainCycle();
  }

  goNext() {
    const lightsState = this.props.lightsState;
    if (lightsState[0]){
      this.setTimeCount(1);
      this.goTo('yellow');
    }
    else if (lightsState[1]) {
      if (this.prevPath === '/green'){
        this.setTimeCount(0);
        this.goTo('red');
      }
      else{
        this.setTimeCount(2);
        this.goTo('green');
      }
    } else if (lightsState[2]) {
      this.setTimeCount(1);
      this.goTo('yellow');
    }
  }
  goTo(path){
    this.props.history.push({
      pathname: path
    });
  }

  onSetClick(val,ind){
    this.times[ind] = val;
  }

  render() {
    const activeColor = this.props.lightsState;
    const doBlink = this.state.timeCount < 3;
    return (
      <div className = 'main-container'>

        <div className='traffic'>
          <div className = 'lights-container'>
            {this.colors.map((el,ind) =>
              <Light key = {el.id}
                isActive = {activeColor[ind]}
                blink = {doBlink && activeColor[ind]}
                color = {el.color}/>
            )}
          </div>
          <div className = 'time'> {this.state.timeCount} </div>
        </div>

        <Navbar times = {this.times} colors = {this.colors}
          onSetClick = {this.onSetClick}
          setTime = {this.setTimeCount} />

      </div>
    );
  }
}

export default Traffic;
