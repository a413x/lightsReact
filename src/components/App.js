import React, { Component } from "react";
import { BrowserRouter as Router, Route, Switch, Redirect } from 'react-router-dom';

import Traffic from './Traffic.js'

class App extends Component {
    constructor(props){
      super(props);
      this.lightsState = [[1,0,0],[0,1,0],[0,0,1]];
    }
    render() {
        return (
          <Router>
            <div>
              <Switch>
                <Route path = '/red' render = {(props) => <Traffic {...props} lightsState = {this.lightsState[0]} />} />
                <Route path = '/yellow' render = {(props) => <Traffic {...props} lightsState = {this.lightsState[1]} />} />
                <Route path = '/green' render = {(props) => <Traffic {...props} lightsState = {this.lightsState[2]} />} />
                <Redirect from='/' to='/red' />
              </Switch>
            </div>
          </Router>
        );
    }
}

export default App;
