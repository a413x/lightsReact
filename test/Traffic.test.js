import React from 'react'
import { shallow, mount } from 'enzyme';

import App from '../src/components/App.js';
import Traffic from '../src/components/Traffic.js';

describe('Tests for Traffic component',()=>{
  const initialLight = 0;
  const colors = [
    {id:0,color:'red'},
    {id:1,color:'yellow'},
    {id:2,color:'green'},
  ];
  const times= [5, 4, 7];
  const initialState = {
    timeCount: times[initialLight],
  };

  let container, traffic, instance;

  beforeEach(()=>{
    jest.useFakeTimers();
    container = mount(<App />);
    traffic = container.find('Traffic');
    instance = traffic.instance();
  });

  it('renders properly',()=>{
    const props = {
      lightsState: [1,0,0],
      location:{pathname:'/red'},
      history:{push:()=>{}}
    }
    const shallowContainer = shallow(
      <Traffic {...props} />
    );
    expect(shallowContainer).toMatchSnapshot();
  });

  it('initialize with initial state',()=>{
    expect(traffic.state()).toEqual(initialState);
  });

  describe('setTimeCount called, should call mainCycle and set timeCount',()=>{
    it('called with 0',()=>{
      jest.spyOn(instance,'mainCycle');
      instance.setTimeCount(0);
      expect(instance.mainCycle).toHaveBeenCalled();
      expect(traffic.state().timeCount).toBe(times[0]);
    });
    it('called with 2',()=>{
      jest.spyOn(instance,'mainCycle');
      instance.setTimeCount(2);
      expect(instance.mainCycle).toHaveBeenCalled();
      expect(traffic.state().timeCount).toBe(times[2]);
    });
  });

  describe('mainCycle function should work correctly',()=>{
    it('initial state',()=>{
      jest.spyOn(instance,'incCount');
      expect(setInterval).toHaveBeenCalled();
      expect(instance.incCount).not.toHaveBeenCalled();
      expect(traffic.state().timeCount).toBe(times[initialLight]);
    });
    it('after 1s delay',()=>{
      jest.spyOn(instance,'incCount');
      jest.advanceTimersByTime(1000);
      expect(instance.incCount).toHaveBeenCalledTimes(1);
      expect(traffic.state().timeCount).toBe(times[initialLight]-1);
    });
  });

  it('incCount should call goNext when counter<=0',()=>{
    jest.spyOn(instance,'goNext');
    traffic.setState({timeCount:0});
    jest.advanceTimersByTime(1000);
    expect(instance.goNext).toHaveBeenCalledTimes(1);
  });

  describe('componentDidUpdate should save previous path',()=>{
    it('pathnames are different',()=>{
      jest.spyOn(instance,'componentDidUpdate');
      const initPath = '/yellow';
      instance.props.history.push({pathname: initPath});
      instance.props.history.push({pathname: '/red'});
      expect(instance.componentDidUpdate).toHaveBeenCalledTimes(2);
      expect(instance.prevPath).toBe(initPath);
    });
    it('pathnames are same',()=>{
      jest.spyOn(instance,'componentDidUpdate');
      const initPath = '/red';
      instance.props.history.push({pathname: initPath});
      instance.props.history.push({pathname: '/green'});
      instance.props.history.push({pathname: '/green'});
      expect(instance.componentDidUpdate).toHaveBeenCalledTimes(3);
      expect(instance.prevPath).toBe(initPath);
    });
  });

  describe('test for goNext function',()=>{
    it('from /red should go to /yellow',()=>{
      instance.props.history.push({pathname: '/red'});
      instance.goNext();
      expect(instance.props.location.pathname).toBe('/yellow');
    });
    it('from /green should go to /yellow',()=>{
      instance.props.history.push({pathname: '/green'});
      instance.goNext();
      expect(instance.props.location.pathname).toBe('/yellow');
    });
    it('from /yellow should go to /green if previous path was /red',()=>{
      instance.props.history.push({pathname: '/red'});
      instance.goNext();
      instance.goNext();
      expect(instance.props.location.pathname).toBe('/green');
    });
    it('from /yellow should go to /red if previous path was /green',()=>{
      instance.props.history.push({pathname: '/green'});
      instance.goNext();
      instance.goNext();
      expect(instance.props.location.pathname).toBe('/red');
    });
  });

  it('overall logic',()=>{
    instance.props.history.push({pathname: '/red'});
    traffic.setState({timeCount:2});
    instance.times = [2,1,2];
    jest.advanceTimersByTime(3000);
    expect(instance.props.location.pathname).toBe('/yellow');
    jest.advanceTimersByTime(2000);
    expect(instance.props.location.pathname).toBe('/green');
    jest.advanceTimersByTime(3000);
    expect(instance.props.location.pathname).toBe('/yellow');
    jest.advanceTimersByTime(2000);
    expect(instance.props.location.pathname).toBe('/red');
    jest.advanceTimersByTime(5000);
    expect(instance.props.location.pathname).toBe('/green');
  });

  afterAll(()=>{
    traffic.setState(initialState);
    container.unmount();
  });
});
