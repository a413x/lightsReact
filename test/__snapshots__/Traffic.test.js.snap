// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Traffic component renders properly 1`] = `
<div
  className="main-container"
>
  <div
    className="traffic"
  >
    <div
      className="lights-container"
    >
      <Light
        blink={false}
        color="red"
        isActive={1}
        key="0"
      />
      <Light
        blink={false}
        color="yellow"
        isActive={0}
        key="1"
      />
      <Light
        blink={false}
        color="green"
        isActive={0}
        key="2"
      />
    </div>
    <div
      className="time"
    >
       
      5
       
    </div>
  </div>
  <Navbar
    colors={
      Array [
        Object {
          "color": "red",
          "id": 0,
        },
        Object {
          "color": "yellow",
          "id": 1,
        },
        Object {
          "color": "green",
          "id": 2,
        },
      ]
    }
    onLinkClick={[Function]}
    onSetClick={[Function]}
    times={
      Array [
        5,
        4,
        7,
      ]
    }
  />
</div>
`;
